sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -76,6 +76,12 @@ abstract class Expression extends TreeNode[Expression] {
       case u: UnresolvedAttribute => PrettyAttribute(u.name)
     }.toString
   }
+
+  def semanticEquals(other: Expression): Boolean = this.getClass == other.getClass &&
+    this.productIterator.zip(other.asInstanceOf[Product].productIterator).forall {
+      case (e1: Expression, e2: Expression) => e1 semanticEquals e2
+      case (i1, i2) => i1 == i2
```

Comment:
The difficulty here is we probably never knows `semanticEquals` in a general way, that's why I said we need to re-implemented for lots of expressions if we added this.